### DISCUSSION TITLE
offline Hedy

### DISCUSSION THREAD

**Discussion by declann** [hedyorg/hedy#5067] [Language Ideas]
Hi!

I was searching about offline, it appears on previous meeting notes: https://discord.com/channels/978580085491785769/1085167289239871538/1085195592050950174

I think never made it over to GitHub, but I don't know about updates after the notes (March 2023).

If it's a desirable feature then maybe we can gather info and ideas here.

There is [hedy-offline](https://github.com/hedyorg/hedy-offline) by Mustafa which is an installable app.

First from meeting notes in Feb:

> Noel took a look at Mustafa’s offline app:
>
> Notes: I took a look at Mustafa's project, and I think it is a pretty cool idea.  Although, I think we can achieve the same/similar with a more gradual approach of converting the Hedy website to a progressive web app. Of course to go completely offline with the PWA approach, we will have to parse/transpile on the client side, likely using Pyodide. 
>
> A benefit of going the PWA route is it reduces/eliminates the drift between the offline and online versions of the app.
>
> We could potentially also create a Language Server, which will help with intellisense, syntax highlighting, debugging, etc, assuming the ACE editor supports LSP.

And also in April (I presume this is related):

> "Noel proposes to walk the parsetree in the browser, or maybe using Pyodide (Python in the browser)"

---

I'm not well versed on Hedy architecture details but I guess someone can point out what I miss with my own thoughts below!

I notice a JS version of Lark: [Lark.js](https://github.com/lark-parser/lark.js) but it hasn't been updated in 10 months. Not sure if it will work with Hedy grammers, or if grammers require Lark- or recent-Lark-only features.

If we can get ASTs using JS, then codegen for Skulpt remains. It **looks like** Noel considered that from his last point above, so essentially converting [hedy.py](https://github.com/hedyorg/hedy/blob/main/hedy.py) to JS? Which would not be trivial and first need to figure how to do the walking.

Alternatively, running the Python for necessary bits in JS through Pyodide (I notice PyScript also but I don't know how that maps to these needs)

Please correct errors and share other ideas!

Declan

**Comment by Felienne** [comment]
Hi Declan! 

Sorry I missed this before because of the hackathon. Having an offline Hedy would be an amazing addition to what we support now, and I would love to have it. Here are a few of my random thoughts about it for safekeeping, not sure if they will be useful to you.

An offline app would be great because it would allow kids to run Hedy if there is no internet access (which is more common than you might think!) It could also be a great first step to a real app. 

The problems that we would have to solve are, in order of my worry about them :) 

* Such a project needs an owner. The team now has a lot of other things going on, and while we really really like this, we can't spend the effort. So we would need a contributor that says: I will make this and then maintain it. Mustafa's project was really nice, but he was a student and now that he is gone, the project is dead

(TBD later...)
